### ISSUE TITLE
addCustomHeader corrupts header value

### ISSUE THREAD

**Issue by clarkk** [PHPMailer/PHPMailer#1854]
`addCustomHeader` corrupts the header value

Original header
`X-My-Header: document/supplier_order block/4431 id/12589 email/alias@mail.com log/21288`

Encoded header (PHPMailer)
`X-My-Header: =?us-ascii?Q?document/supplier=5Forder_block/4431_id/12589_email/alias@mail?=
 =?us-ascii?Q?.com_log/21288?=`

Decoded header
`X-My-Header: document/supplier_order_block/4431_id/12589_email/alias@mail.com_log/21288`

The Original header and the decoded header are not identical. Spaces are translated to _ (underscore)

The header is decoded with `mb_decode_mimeheader()`

**Comment by Synchro** [comment]
That header is correctly Q-encoded according to [RFC2047 section 4.2(2)](https://tools.ietf.org/html/rfc2047#section-4.2), so this looks like a bug in `mb_mimedecode_header`.

**Comment by Synchro** [comment]
[Reported as a PHP bug](https://bugs.php.net/bug.php?id=78617).

**Comment by clarkk** [comment]
`iconv` can also work differently depending on the system, so I try not to use them

**Comment by clarkk** [comment]
but it would still be nice to pass an extra bool argument to `addCustomHeader` weather to disable the encoding

**Comment by clarkk** [comment]
or at least check the value before encoding it if it contain "invalid" chars

**Comment by Synchro** [comment]
I think that's a pretty terrible idea. If you don't encode the headers when it's needed, you will have corrupted headers. I'm not doing that. You can always subclass if you want to break things on purpose.

**Comment by clarkk** [comment]
What about checking if the header value contains any illegal chars before encoding?

**Comment by Synchro** [comment]
Have you tried reading [the source](https://github.com/PHPMailer/PHPMailer/blob/master/src/PHPMailer.php#L3212)? That's exactly what it does.

**Comment by clarkk** [comment]
I have.. :)

The example `X-My-Header: document/supplier_order block/4431 id/12589 email/alias@mail.com log/21288` contains no illegal chars but it is encoded anyway.. It may be a bit long but then you only have to break the header - not encoding it

**Comment by Synchro** [comment]
Look at the encoder - it will apply Q encoding if the line contains non-ascii chars, or if its length means it needs folding (even if it only contains ascii). The reason for that is that the PHP `mail()` function can corrupt lines longer than 65 chars, even though SMTP is supposed to allow at least 76 (and up to 998). Q encoding the header value means that it can be folded safely even if the header value does not contain any spaces; it's not possible to fold a header without spaces except by using RFC2047 encoding. Look at the discussion around it in #1525, and the fix for it in #1825.

In short, if you don't encode the header correctly, you're asking for all kinds of other trouble. Avoiding encoding it on the basis that your decoder is broken isn't really a good reason.

**Comment by clarkk** [comment]
Ok, your the expert :)

The headers are just alot prettier when not encoded

**Comment by Synchro** [comment]
🤷‍♂️ We're not here for pretty, we're here for functional and compliant!
